File: sdk/js/pools.js
import StorageManagerAuthError from './helpers';

export default class Pool {
    constructor(mgr, name) {
        this.mgr = mgr;
        this.name = name;
    }

    static async create(mgr, name, distribute_groups, opts) {
        return await mgr.httpPost(`/api/v1/pools`, {
            name: name,
            distribute_groups: distribute_groups,
            no_start: opts["no_start"] !== undefined ? opts["no_start"] : false,
            distribute: opts["distribute"] !== undefined ? opts["distribute"] : false,
            pool_id: opts["pool_id"] !== undefined ? opts["pool_id"] : "",
            auto_add_nodes: opts["auto_add_nodes"] !== undefined ? opts["auto_add_nodes"] : false,
            options: opts["options"] !== undefined ? opts["options"] : {},
        })
    }

    static async list(mgr, state=false) {
        return await mgr.httpGet(`/api/v1/pools?state=${state ? 1 : 0}`)
    }

    async get(state=false) {
        return await this.mgr.httpGet(
            `/api/v1/pools/${this.name}?state=${state ? 1 : 0}`
        )
    }

    async start() {
        return await this.mgr.httpPost(
            `/api/v1/pools/${this.name}/start`, {}
        )
    }

    async stop() {
        return await this.mgr.httpPost(
            `/api/v1/pools/${this.name}/stop`, {}
        )
    }

    async rename(newName) {
        return this.mgr.httpPost(
            `/api/v1/pools/${this.name}/rename`,
            {new_name: newName}
        )
    }
    
    async delete() {
        return this.mgr.httpDelete(
            `/api/v1/pools/${this.name}`
        )
    }
}

File: sdk/python/kadalu_storage/pools.py
# noqa # pylint: disable=missing-module-docstring

from kadalu_storage.helpers import response_object_or_error


class Pool:
    # noqa # pylint: disable=missing-class-docstring
    def __init__(self, mgr, name):
        self.mgr = mgr
        self.name = name

    @classmethod
    def list(cls, mgr, state = False):
        # noqa # pylint: disable=missing-function-docstring
        resp = mgr.http_get(f'{mgr.url}/api/v1/pools?state={1 if state else 0}')
        return response_object_or_error("Pool", resp, 200)

    @classmethod
    def create(cls, mgr, name, distribute_groups, options=None):
        # noqa # pylint: disable=missing-function-docstring
        # noqa # pylint: disable=too-many-arguments
        options = {} if options is None else options
        resp = mgr.http_post(
            f"{mgr.url}/api/v1/pools",
            {
                "name": name,
                "distribute_groups": distribute_groups,
                "no_start": options.get("no_start", False),
                "distribute": options.get("distribute", False),
                "pool_id": options.get("pool_id", ""),
                "auto_add_nodes": options.get("auto_add_nodes", False),
                "options": options.get("options", {})
            }
        )
        return response_object_or_error("Pool", resp, 201)

    def start(self):
        """
        == Start a Kadalu Storage Pool

        Start a Kadalu Storage Pool

        Example:

        [source,python]
        ----
        from kadalu_storage import StorageManager

        mgr = StorageManager("http://localhost:3000")

        mgr.pool("pool1").start()
        ----
        """
        resp = self.mgr.http_post(
            f"{self.mgr.url}/api/v1/pools/{self.name}/start"
        )
        return response_object_or_error("Pool", resp, 200)

    def stop(self):
        """
        == Stop a Kadalu Storage Pool

        Stop a Kadalu Storage Pool

        Example:

        [source,python]
        ----
        from kadalu_storage import StorageManager

        mgr = StorageManager("http://localhost:3000")

        mgr.pool("pool1").stop()
        ----
        """
        resp = self.mgr.http_post(
            f"{self.mgr.url}/api/v1/pools/{self.name}/stop"
        )
        return response_object_or_error("Pool", resp, 200)

    def delete(self):
        """
        == Delete a Kadalu Storage Pool

        Delete a Kadalu Storage Pool

        Example:

        [source,python]
        ----
        from kadalu_storage import StorageManager

        mgr = StorageManager("http://localhost:3000")

        mgr.pool("pool1").delete()
        ----
        """
        url = f"{self.mgr.url}/api/v1/pools/{self.name}"
        resp = self.mgr.http_delete(url)
        return response_object_or_error("Pool", resp, 204)
